### PR TITLE
Add `StreamMuxer::flush`.

### LINKED ISSUE #493: No way to flush a muxer after shutting down the inbound or outbound channels
Right now the only explicit way to flush a `StreamMuxer` is to flush one of its substreams.

An implementation of `StreamMuxer` may require sending a packet of data to tell the remote about the closing. If the underlying stream is not ready at the time when `close_substream()` is called, then there is no way to do that other than waiting for the user to perform another action on that muxer.

### PR DESCRIPTION
**Pull request by twittner**
~~Currently uses the `flush` branch of yamux. Should be moved back to master once https://github.com/paritytech/yamux/pull/18 is accepted.~~

Closes #493.

### CODE REVIEW

File: core/src/muxing.rs [hunk 1]
```diff
@@ -60,46 +60,35 @@ pub trait StreamMuxer {
...
     /// If `WouldBlock` is returned, then the current task will be notified once the substream
     /// is ready to be read, similar to the API of `AsyncRead`.
     /// However, for each individual substream, only the latest task that was used to call this
     /// method may be notified.
-    fn read_substream(
-        &self,
-        substream: &mut Self::Substream,
-        buf: &mut [u8],
-    ) -> Result<usize, IoError>;
+    fn read_substream(&self, s: &mut Self::Substream, buf: &mut [u8]) -> Result<usize, IoError>;
```
Comment (tomaka) [inline core/src/muxing.rs]: Why rename `substream` to `s`? This is less descriptive.

**Comment by tomaka** [comment]
I think we should be more precise in what `flush()` does, ie. does it flush the pending data on all the substreams? Or does it only flush the command that tells the remote that we're closing the inbound or outbound channel?
I would go for the latter, but I'm a bit undecided on this question.

My reasoning is by taking a look at the operations that write on the socket:

- Opening a substream. If opening a substream requires a response from the remote, I expect `poll_outbound` to flush at least the packet that request a new substream.
- Responding to a new substream request ; I would expect `poll_inbound()` to handle flushing.
- Writing on an existing substream. We have `flush_substream()`.
- Closing a substream. We have `shutdown_substream()` that I think should flush.
- Responding to an incoming message, for example saying that we're overloaded. It is debatable whether `read_substream()` should flush that, as I guess it depends on the nature of the message. However worst case scenario I would expect `flush_substream()` to handle flushing, and not a separate `flush()` method.
- Closing the inbound or outbound channel, which is the only missing thing.

File: core/src/muxing.rs (hunk 1 above)
Comment (twittner) [reply core/src/muxing.rs]: Given that it has `Substream` in its type right next to it and does not use the variable name since there is no method body I felt that having the method declaration on a single line is better.

**Comment by twittner** [comment]
I would expect the `StreamMuxer` impl to handle required flushing itself where it is required. For users I think the important semantics are (a) being able to flush a particular substream (cf. `flush_substream`) and (b) being able to flush all of it, i.e. the data of all substreams (cf. `flush`). 

If  `close_inbound` and `close_outbound` should be given the chance to send a message, their signatures should change to `fn close_outbound(&self) -> Poll<(), io::Error>` and `fn close_inbound(&self) -> Poll<(), io::Error>`.

**Comment by tomaka** [comment]
> (b) being able to flush all of it, i.e. the data of all substreams

What would be the point of doing that?
When it comes to flushing, it's not that the user *wants* to flush, it's that they *need* to flush, and therefore I think you need to indicate when it is required to flush.

**Comment by twittner** [comment]
True. So the suggested change to  `close_inbound` and `close_outbound` should be sufficient then.

**Comment by tomaka** [comment]
> So the suggested change to close_inbound and close_outbound should be sufficient then.

The "problem" is that if you want to close both, then you'd flush twice. Hence the separate method.

**Comment by twittner** [comment]
I think the close_* methods are called very infrequently if at all and having them return `Poll` leaves no room for error w.r.t. flushing.

> On the other than the user may want to shut down say 5 substreams and close the outbound side, and then only flush everything.

I am confused. To me this use case would have been supported by `StreamMuxer::flush`. I (maybe by mistake) read your previous comments as a rejection of `StreamMuxer::flush` in preference for fine-grained implicit flushing by the various existing methods. Of course, a combination of those methods may incur the overhead of multiple implicit flushes.

**Comment by tomaka** [comment]
> I (maybe by mistake) read your previous comments as a rejection of StreamMuxer::flush in preference for fine-grained implicit flushing by the various existing methods.

I'm not really trying to argue in favour of one solution, I'm still very conflicted.

**Comment by twittner** [comment]
I changed the `StreamMuxer` trait as follows:

1. `read_substream`, `write_substream` and `flush_substream` now return `Poll` instead of `Result`.
2. A new `Shutdown` enum allows for half-closing of substreams and is used in `shutdown_substream`.
3. `close_inbound` and `close_outbound` have been merged into `shutdown` which takes a `Shutdown` parameter to allow closing only one direction.

W.r.t. flushing the general idea is that normal use drains buffers over time. Shutting down a substream does not imply flushing, so can be followed  by `flush_substream` or (if multiple substreams are to be shut down) a single `flush_all`. Shutting down the muxer itself proceeds likewise, i.e. `shutdown` followed by `flush_all`.

File: core/src/muxing.rs [hunk 2]
```diff
@@ -61,66 +71,61 @@ pub trait StreamMuxer {
...
 
+    /// Shutdown this `StreamMuxer`.
+    ///
     /// If supported, sends a hint to the remote that we may no longer open any further outbound
-    /// substream. Calling `poll_outbound` afterwards may or may not produce `None`.
-    fn close_outbound(&self);
+    /// or inbound substream. Calling `poll_outbound` or `poll_inbound` afterwards may or may not
+    /// produce `None`.
+    ///
+    /// Shutting dow the muxer does not imply `flush_all`. If you want to make sure that the
```
Comment (tomaka) [inline core/src/muxing.rs]: s/dow/down

File: core/src/nodes/node.rs [hunk 3]
```diff
@@ -310,11 +310,18 @@ where
         for (_, outbound) in self.outbound_substreams.drain() {
             self.muxer.destroy_outbound(outbound);
         }
-        if !self.inbound_finished {
-            self.muxer.close_inbound();
-        }
-        if !self.outbound_finished {
-            self.muxer.close_outbound();
+        // TODO: Maybe the shutdown logic should not be part of the destructor?
```
Comment (tomaka) [inline core/src/nodes/node.rs]: The node should definitely get a `shutdown` method that cleanly shuts down the muxer.
This can be in a separate PR though.

File: muxers/mplex/src/lib.rs [hunk 4]
```diff
@@ -475,50 +475,51 @@ where C: AsyncRead + AsyncWrite
...
+            endpoint: sub.endpoint,
         };
 
         let mut inner = self.inner.lock();
         poll_send(&mut inner, elem)
     }
 
     fn destroy_substream(&self, mut substream: Self::Substream) {
-        let _ = self.shutdown_substream(&mut substream);        // TODO: this doesn't necessarily send the close message
+        let _ = self.shutdown_substream(&mut substream, Shutdown::All);        // TODO: this doesn't necessarily send the close message
```
Comment (tomaka) [inline muxers/mplex/src/lib.rs]: I wonder if we can remove this, or if we don't we should remove the TODO.

**Review by tomaka — CHANGES_REQUESTED** [review]

File: muxers/mplex/src/lib.rs (hunk 4 above)
Comment (twittner) [reply muxers/mplex/src/lib.rs]: I think we can remove it.

**Review by tomaka — APPROVED** [review]

**Review by tomaka — CHANGES_REQUESTED** [review]
Oops, cancel that approval.

**Comment by tomaka** [comment]
I think mplex and yamux should call `close` on the underlying stream when it is shut down.
If you don't want to do that, say so and open an issue instead.

**Comment by twittner** [comment]
`mplex` shuts down the connection, for yamux see https://github.com/paritytech/yamux/issues/19.

**Review by tomaka — APPROVED** [review]
